### ISSUE TITLE
First `.request()` call slow

### ISSUE THREAD

**Issue by mkarg** [eclipse-ee4j/jersey#5281]
Using Jersey 3.1.1 (JAX-RS Client API) I noticed that the very first call to `.request()` needs between 500ms and 1000ms, while subsequent calls to `.request()` need 0 ms.

While this is not a problem when the Client API is called on a server (hence when a service calls into another service), as typically a lot of subsequent calls will be performed, it is a problem when the Client API is called from a CLI command. In my case, the command will perform *only* a single request!

What can an application programmer do to reduce this time?

**Comment by jansupol** [comment]
Yes, the first call is slow. There is a multiple reasons, starting with JDK Warmup, ending with Jersey client runtime initialization time.

While I haven't tried any possibility on my own, on the of my head there can be the following options:  

1. Warming up JDK as described by many blogs, for instance [Baeldung](https://www.baeldung.com/java-jvm-warmup)
2. Native image, see https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/graalvm-native-image.html
3. Trying some JDK features, such as project [Leyden](https://blogs.oracle.com/javamagazine/post/java-leyden-static-images). While it seems to be a think for the future, there might be a similar initiative by [OpenJ9](https://danheidinga.github.io/Everyone_wants_fast_startup/)

**Comment by mkarg** [comment]
Thank you Jan. These steps are pretty clear to me, but in fact do not help (we have some of them in place already) to speed up *this particular* call a lot (up to *this particukar* call, everything happens in few Milliseconds, even in interpreted mode). The reason *here* is that this is *not* the typical JDK warmup problem! Even with tricks like these, the difference between the first and second call is *huge* (seconds, not milliseconds). And even in pure interpreted mode it is unclear why Jersey needs one second just for this single method. We assume Jersey waits for something, but for what? Apparently it seems Jersey itself is doing a lot of things here that are actually *not needed* in our case, for example searching for Activation API, so possibly a lot of these actions could get skipped if there just would be a property to tell Jersey...? I think a "real" solution is to make parts of the bootstrapping *optional*. That would bringt more benefit that running unnecessary things in higher speed. WDYT?

**Comment by jansupol** [comment]
Yep, it would have been nice, would it not?

Each Jersey Client starts an instance of HK2, registers all the providers there, creates the instances, searches for services, as you named...That takes time. 

I tried to drop any injection support in the client to speedup the requests, and the startup time is about halved. If your client does not use injections, you can drop HK2 and use (in 2.39.1) [inject-less client](https://github.com/eclipse-ee4j/jersey/tree/master/incubator/injectless-client).
```
        <dependency>
            <groupId>org.glassfish.jersey.incubator</groupId>
            <artifactId>jersey-injectless-client</artifactId>
            <version>2.39.1</version>
        </dependency>
```
regarding the activation API & al. search, this can be turned off by `.property(CommonProperties.PROVIDER_DEFAULT_DISABLE, "ALL")`.

**Comment by jansupol** [comment]
It's going to be in the next 3.x, too.

**Comment by mkarg** [comment]
@jansupol Thank you so much for these valuable tipps! You are awesome! :-)

For those who are interested in some actual numbers (measured using Jersey Client 2.39.1):

* Jersey 2.39.q HK2 (baseline): 100% (1469,2 ms)
* `jersey-injectless-client` instead of HK2: 18,09% performance improvement
* `jersey.config.disableDefaultProvider = ALL`: 19,51% performance improvement
* Both tricks applied at the same time: **32,97%** performance improvement

This looks very promising and I really appreciate these tuning options!

Having said that, I'd like to encourage everybody to contribute more performance optimizations, as a program never can be fast enough! :-)

**Comment by jansupol** [comment]
@mkarg Thank you for sharing the results, they are really interesting.
